README
NAME
    WWW::CPAN - CPAN as a web service

SYNOPSIS
      use WWW::CPAN;

      $meta = WWW::CPAN->fetch_distmeta('Error');

      $cpan = WWW::CPAN->new;
      $meta = $cpan->fetch_distmeta('Error');
      $meta = $cpan->fetch_distmeta({ dist => 'Error' });
      $meta = $cpan->fetch_distmeta({ dist => 'Error-0.13', });
      $meta = $cpan->fetch_distmeta({ dist => 'Error', version => '0.13' }); # same as above
      $meta = $cpan->fetch_distmeta({ dist => 'Error', author => 'GBARR', format => 'json' });

      $result = $cpan->search('WWW-CPAN');
      $result = $cpan->search({ query => 'WWW-CPAN', mode => 'dist' });

DESCRIPTION
    Once upon a time, Schwern waved his hands and Graham implemented a
    mechanism to get the meta data from CPAN distributions at
    search.cpan.org.

      http://use.perl.org/~schwern/journal/35203

      http://www.mail-archive.com/module-build@perl.org/msg01157.html

    This module opportunistically implemented a client for this capability.
    One day, hopefully this may be useful.

    A further addition to `WWW::CPAN' API was the `search' method to fetch
    query data from search.cpan.org. This is done by issuing HTTP requests
    like

      http://search.cpan.org/search?query=Romanize&mode=all&format=xml

METHODS
    fetch_distmeta
          $meta = $cpan->fetch_distmeta($text);
          $meta = $cpan->fetch_distmeta(\%params);

        Fetches the meta data (META.yml) of a CPAN distribution.

        Supported parameters:

          dist    - the distribution to look for
          author  - the distribution author
          version - the distribution version

    search
          $res = $cpan->search($text);
          $res = $cpan->search(\%params);

        Performs a query at CPAN and returns it as a (possibly partial)
        result set. The structure of the result is illustrated by the
        example below.

          # result of   WWW::CPAN->search({ query => 'CPAN', mode => 'dist', n => 2 })
          {
            dist    => [
                         {
                           author  => { link => 'http://search.cpan.org/~andk/' },
                           description
                                   => 'query, download and build perl modules from CPAN sites ',
                           link    => 'http://search.cpan.org/author/ANDK/CPAN-1.9205/',
                           name    => 'CPAN',
                           released
                                   => '11th November 2007',
                           version => 1.9205
                         },
                         {
                           author  => { link => 'http://search.cpan.org/~dons/' },
                           description
                                   => 'Download and install CPAN modules upon first use',
                           link    => 'http://search.cpan.org/author/DONS/CPAN-AutoINC-0.01/',
                           name    => 'CPAN-AutoINC',
                           released
                                   => '22nd June 2004',
                           version => 0.01
                         }
                       ],
            matches => 108,
            start   => 1
         };

        The known parameters of the query are

          query  - what to look for
          mode   - all | module | dist | author
          n      - page size (should be <= 100)
          s      - start

        The maximum page size (`n') is 100. Larger values are reset into
        100.

    query
          $res = $cpan->query($text);
          $res = $cpan->query(\%params);

        This is a method alias to search. It works just the same.

SEE ALSO
      cpanq
      App::WWW::CPAN

    cpanq is a script which illustrates the retrieval capabilities of this
    module from the command line.

BUGS
    In the distribution test t/03meta.t, when using
    `WWW::CPAN->fetch_distmeta' instead of `WWW::CPAN->new->fetch_distmeta',
    the script crashes (look for the "FIXME" item). Should be fixed in the
    next releases (after I understood what's going on).

    Please report bugs via CPAN RT
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-CPAN or mailto:.

AUTHORS
    Adriano R. Ferreira, <ferreira@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2007, 2008 by Adriano R. Ferreira

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for WWW-CPAN.

0.011   Sat Mar  1 2008
        - try Cache::FileCache rather than Cache::File:
          it seems to work on Windows (at least more than
          Cache::File does)

0.010   Sat Mar  1 2008
        - experiment LWP::UserAgent::Determined to get
          rid of some failures in live test
        - dependency on LWP::UserAgent replaced by 
          L::UA::Determined for now
        - nope, it won't help: search.cpan.org is playing
          tricks on me - the META.json of CPAN
          is auto-generated sometimes instead of
          being the one extracted from the current dist

0.009   Thu Feb 21 2008
        - 'search' is now the preferred name of 'query'
          in WWW:CPAN API
        - in cpanq, 'search' is now the preferred name
          of 'query' command
        - in cpanq, 'distmeta' is a synonym of 'meta'
          command
        - cpanq now produces a usage message when
          no arguments are given
        - Pod::Usage is now a prereq

0.008   Tue Feb 19 2008
        - update dependency: Class::Constructor::Factory 
          renamed to Class::Lego::Constructor
        - usage pattern of Class::Lego::Constructor updated

0.007   Tue Jan 29 2008
        - minor internal change to load JSON/XML
        - cpanq does not display "$VAR1 = undef"
          on errors anymore
        - cpanq now uses a cache
        - the live test t/03meta.t announces itself
        - t/03meta.t can be skipped (given an argument
          --no-live-tests)
        - added t/98pod-coverage.t and t/99pod.t
        - added forgotten dependency on URI 
          (thanks, CPANTS!)

0.006    Sun Jan 27 2008
        - more tests (including a live test)

0.005    Sat Jan 26 2008
        - included a demo script: cpanq

0.004    Mon Jan 14 2008
        - query() added to the API
        - more POD
        - new dependency on XML::Simple

0.003    Wed Jan  9 2008
        - new dependency on Class::Lego::Myself

0.002    Thu Dec 27 2007
        - carp and return undef rather than die
        - more POD
        - Class::Constructor::Factory is used for generating
          the constructor with field default values

0.001    Wed Dec 26 2007
        - first CPAN release
        - new() and fetch_distmeta() is the whole API


t/lib/Devel/CheckNet.pm

package Devel::CheckNet;

use strict;
use warnings;

our $VERSION = '0.001';

use LWP::UserAgent ();

sub can_http {
  my $self = shift;
  my $url = shift;

  my %ua_options = (
    agent   => 'devel-checknet/' . $VERSION,
    timeout => 10,
    requests_redirectable => [],
    parse_head => 0,
  );
  my $ua = LWP::UserAgent->new();
  my $r = $ua->head( $url );
  return ( !$r->is_error ) ? 1 : 0;
}

1;

__END__

=head1 NAME

Devel::CheckNet - 

=head1 SYNOPSIS

  use Devel::CheckNet ();

  $ok = Devel::CheckNet->can_http( 'http://search.cpan.org/' );



lib/WWW/CPAN.pm

package WWW::CPAN;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.011'; 

use Class::Lego::Constructor 0.004 ();
use parent qw( Class::Accessor Class::Lego::Constructor );

my $FIELDS = {
  host     => 'search.cpan.org',
  ua       => sub { # default useragent
                my %options = ( agent => 'www-cpan/' . $VERSION, );
#                require LWP::UserAgent;
#                return LWP::UserAgent->new( %options );
                require LWP::UserAgent::Determined;
                return LWP::UserAgent::Determined->new( %options );
              },
  j_loader => sub { # json loader sub
                require JSON::Any;
                JSON::Any->import; # XXX JSON::Any needs this
                my $j = JSON::Any->new;
                return sub { $j->Load(shift); }
              },
  x_loader => sub { # xml loader sub
                require XML::Simple;
                my %options = (
                  ForceArray => [qw( module dist match )],
                  KeyAttr    => [],
                );
                my $x = XML::Simple->new( %options );
                return sub { $x->XMLin(shift); }
              },
};

__PACKAGE__->mk_constructor0( $FIELDS );
__PACKAGE__->mk_accessors( keys %$FIELDS );

use Class::Lego::Myself;
__PACKAGE__->give_my_self;

use Carp qw( carp );

sub _build_distmeta_uri {
  my $self = shift;
  my $params = shift;

  if ( ! ref $params ) {
    $params = { dist => $params };
  }
  require URI;
  my $uri = URI->new();
  $uri->scheme('http');
  $uri->authority( $self->host );
  my @path = qw( meta );
  if ( $params->{author} ) {
    push @path, $params->{author};
  }

  my $dist = $params->{dist};
  if ( $params->{version} ) {
    $dist .= '-' . $params->{version};
  }
  push @path, $dist;

  push @path, 'META.json'; # XXX support YAML as well
  $uri->path_segments(@path);

  return $uri;
}

sub fetch_distmeta {
  (my $self, @_) = &find_my_self;
  my $uri = $self->_build_distmeta_uri(@_);
  my $r = $self->ua->get($uri);
  if ( $r->is_success ) {
    return $self->j_loader->( $r->content );
  } else {
    carp $r->status_line; # FIXME needs more convincing error handling
    return;
  }
}

# http://search.cpan.org/search?query=Archive&mode=all&format=xml
sub _build_query_uri {
  my $self = shift;
  my $params = shift;

  if ( ! ref $params ) {
    $params = { query => $params, mode => 'all', format => 'xml', };
  }
  require URI;
  my $uri = URI->new();
  $uri->scheme('http');
  $uri->authority( $self->host );
  my @path = qw( search );
  $uri->path_segments(@path);

  $params->{mode}   ||= 'all';
  $params->{format} ||= 'xml';
  $uri->query_form( $params );

  return $uri;
}
# other params: s (start), n (page size, should be <= 100)

sub _basic_query {
  my $self = shift;
  my $uri = $self->_build_query_uri(@_);
  my $r = $self->ua->get($uri);
  if ( $r->is_success ) {
    return $self->x_loader->( $r->content );
  } else {
    carp $r->status_line; # FIXME needs more convincing error handling
    return;
  }
}

sub search {
  my $self = &find_my_self;
  return $self->_basic_query(@_);
}
# TODO fetch the entire result by default

# &query is an alias to &search (see Method::Alias for the rationale)
sub query { 
  goto &{ $_[0]->can('search') }; 
}

"I didn't do it! -- Bart Simpson";


Makefile.PL

use 5.006;
use ExtUtils::MakeMaker 6.31; # XXX yes, that's not a nice thing to do

WriteMakefile(
  NAME         => 'WWW-CPAN',
  VERSION_FROM => 'lib/WWW/CPAN.pm',

  PREREQ_PM => {
    'strict'     => 0,
    'warnings'   => 0,

#    'LWP::UserAgent'   => 0,
    'LWP::UserAgent::Determined' => 0,
    'URI'              => 0,
    'parent'           => 0,
    'Class::Accessor'  => 0,
    'Class::Lego::Constructor' => 0.004,
    'Class::Lego::Myself'         => 0.001,
    'Carp'             => 0,

    'JSON::Any'        => 0,
    'XML::Simple'      => 0,
    # required by bin/cpanq
    'Data::Dump::Streamer' => 0,
    'Cache::FileCache'     => 0,
    'File::HomeDir'        => 0,
    'Path::Class'          => 0,
    'Pod::Usage'           => 0,

    'Test::More'       => 0,

  },
  # recommends => { 
  #       'Test::Script' => 0, 
  #}
  EXE_FILES     => [ 'bin/cpanq' ],
  LICENSE       => 'perl',
  AUTHOR        => 'Adriano R. Ferreira <ferreira@cpan.org>',
  ABSTRACT_FROM => 'lib/WWW/CPAN.pod',

);
